src/test/java/at/ac/tuwien/kr/alpha/api/query/AnswerSetQueryTest.java

```diff
@@ -0,0 +1,116 @@
+	}
+
+	@Test
+	public void matchSymbolicConstant() {
+		AnswerSetBuilder bld = new AnswerSetBuilder();
+		bld.predicate("p")
+				.symbolicInstance("a")
+				.instance("a");
+		AnswerSet as = bld.build();
+		AnswerSetQuery constantQuery = AnswerSetQuery
+				.forPredicate(Predicate.getInstance("p", 1))
+				.withConstantEquals(0, "a");
+		List<Atom> queryResult = as.query(constantQuery);
+		Assert.assertEquals(1, queryResult.size());
+	}
+
+	@Test
+	public void matchString() {
+		AnswerSetBuilder bld = new AnswerSetBuilder();
+		bld.predicate("p")
+				.symbolicInstance("a")
+				.instance("a");
+		AnswerSet as = bld.build();
+		AnswerSetQuery stringQuery = AnswerSetQuery
+				.forPredicate(Predicate.getInstance("p", 1))
+				.withStringEquals(0, "a");
+		List<Atom> queryResult = as.query(stringQuery);
+		Assert.assertEquals(1, queryResult.size());
+	}
+
+	@Test
+	public void matchEvenIntegers() {
+		AnswerSetBuilder bld = new AnswerSetBuilder();
+		bld.predicate("p")
+				.instance(1).instance(2).instance(3).instance(4).instance(5)
+				.instance("bla").symbolicInstance("blubb");
+		AnswerSet as = bld.build();
+		java.util.function.Predicate<Term> isInteger = (term) -> {
+			if (!(term instanceof ConstantTerm<?>)) {
+				return false;
+			}
+			String strValue = ((ConstantTerm<?>) term).getObject().toString();
+			return strValue.matches("[0-9]+");
+		};
+		AnswerSetQuery evenIntegers = AnswerSetQuery.forPredicate(Predicate.getInstance("p", 1))
+				.withFilter(0, isInteger.and(
+						(term) -> Integer.valueOf(((ConstantTerm<?>) term).getObject().toString()) % 2 == 0));
+		List<Atom> queryResult = as.query(evenIntegers);
+		Assert.assertEquals(2, queryResult.size());
+		for (Atom atom : queryResult) {
+			ConstantTerm<?> term = (ConstantTerm<?>) atom.getTerms().get(0);
+			Assert.assertTrue(Integer.valueOf(term.getObject().toString()) % 2 == 0);
+		}
+	}
+
+	@Test
+	public void matchXWithFuncTerm() {
+		Predicate p = Predicate.getInstance("p", 2);
+		Atom a1 = new BasicAtom(p, ConstantTerm.getSymbolicInstance("x"), FunctionTerm.getInstance("f", ConstantTerm.getSymbolicInstance("x")));
+		Atom a2 = new BasicAtom(p, ConstantTerm.getSymbolicInstance("y"), FunctionTerm.getInstance("f", ConstantTerm.getSymbolicInstance("y")));
+		Atom a3 = new BasicAtom(p, ConstantTerm.getSymbolicInstance("y"), FunctionTerm.getInstance("f", ConstantTerm.getSymbolicInstance("x")));
+		Atom a4 = new BasicAtom(p, ConstantTerm.getSymbolicInstance("x"), FunctionTerm.getInstance("f", ConstantTerm.getSymbolicInstance("y")));
+		Atom a5 = new BasicAtom(p, ConstantTerm.getSymbolicInstance("x"), ConstantTerm.getSymbolicInstance("f"));
+		SortedSet<Predicate> predicates = new TreeSet<>();
+		predicates.add(p);
+		Map<Predicate, SortedSet<Atom>> instances = new HashMap<>();
+		SortedSet<Atom> ps = new TreeSet<>();
+		ps.add(a1);
+		ps.add(a2);
+		ps.add(a3);
+		ps.add(a4);
+		ps.add(a5);
+		instances.put(p, ps);
+		AnswerSet as = new BasicAnswerSet(predicates, instances);
```

Comment:
Trying to reconstruct my train of thought from a few months ago here, but I think I decided that, since we don't have builders for function terms, I wouldn't get to "streamlined" code anyway.

Reply:
True, what would be needed is a method `AnswerSetBuilder#instance(Term... terms)` that directly accepts terms -- or alternatively a way to parse function terms, but that might be over the top.

With the instance method accepting a list of `Term`, you could then just write `asb.instance(ConstantTerm.getSymbolicInstance("x"), FunctionTerm.getInstance("f", ConstantTerm.getSymbolicInstance("x"));` five times, which would already look much better.